### PR TITLE
Change to use an unified constant for max media attachments per status

### PR DESCRIPTION
**Pull request by ileodo**
Currently, there are multiple places referring a hard coded magic number 4 as the max media attachment per status. This PR is creating a constant in MediaAttaments Mode to unifying the definition.

both JS and ruby will use the same definition.

### CODE REVIEW

**Review by mjankowski — APPROVED** [review]
Looks good overall. Some followup ideas to add to this PR or afterwards:

- The `PostStatusService` spec does check the 4 attachment limit -- this could be update to stub_const this new constant limit and simplify the spec (or at minimum, change the hard-coded `4` there to just reference the constant).
- The `UpdateStatusService` is missing this spec coverage - could be added in similar way.

**Comment by ileodo** [comment]
Thanks @mjankowski could you please help to enable the CI action for this PR?
Also do you want me to follow up on the two points? I'm not super familiar with Ruby on Rails, but I can give a try (it might take longer)

**Review by renchap — DISMISSED** [review]
Looks good!

Did you try locally to increase the value locally and check that both the UI and the API support the new value?

**Comment by ileodo** [comment]
> Looks good!
> 
> Did you try locally to increase the value locally and check that both the UI and the API support the new value?

thanks!  yes, I did tested it locally with changing the const number.

**Comment by renchap** [comment]
It looks like this is breaking the end-to-end tests in the admin panel, specifically this one: https://github.com/mastodon/mastodon/blob/main/spec/system/report_interface_spec.rb#L23

Can you please look at this and fix it?

**Comment by ClearlyClaire** [comment]
I think the admin panel failure is because the Media Gallery gets mounted out of a redux context there.

**Comment by ileodo** [comment]
looks like there is some issue with the current main branch, under development mode (in docker), admin@localhost is not approved.. so I can't replicate and test the change, any idea about the "approved" issue?

**Comment by renchap** [comment]
You can approve it from the CLI using `tootctl modify <username> --approved`

**Comment by ileodo** [comment]
@ClearlyClaire Could you please help with fixing the media gallery in the report view? I'm not very familiar with how redux is used in mastodon, but definitely want to push for this PR. Thanks.

**Comment by renchap** [comment]
The Redux store is not initialised at all when we render independent React components in a page (in your case, `<MediaContainer>`).

A way to solve this is to properly initialise Redux when mounting a `MediaContainer` so it has access to the Redux state. This is probably overkill in this specific case as you are only interested in one specific field from the `instance` endpoint.
I did this here, but I am not sure this is the best way to handle your issue: https://github.com/renchap/mastodon/commit/534ff0cba98ee66f3d8c3fd421bdb6f4f1fa78b0

Another way to do this would be to somehow detect if we are in a Redux context, and if not, fetch this value and pass it to the component, but I do not think we do this in any other place?

**Comment by ileodo** [comment]
Thanks @renchap! given MediaContainer is used in a few different place, I do think having a consistent way to initialise Redux store is preferred. should we just start a new PR  from here ? https://github.com/mastodon/mastodon/compare/main...renchap:mastodon:pr-29073

Thanks

**Comment by ileodo** [comment]
great. CI passed!

**Comment by ileodo** [comment]
bouncing this up.

File: app/serializers/rest/instance_serializer.rb
Comment (mjankowski) [inline app/serializers/rest/instance_serializer.rb:62]: The ruby side of this has been handled in https://github.com/mastodon/mastodon/pull/30433 - including exposing the value via api serializer change. This could rebase against that and be just (I think?) the JS-side changes?

Comment (ileodo) [reply app/serializers/rest/instance_serializer.rb:62]: Okay, will update, didn't know there is another PR working on the exactly same thing.

Comment (mjankowski) [reply app/serializers/rest/instance_serializer.rb:62]: Cool, thanks -- and yes, not trying to make dupe PRs, but despite having comment on this in Feb, I forgot it existed and opened that other one while noticing the issue.

Comment (ileodo) [reply app/serializers/rest/instance_serializer.rb]: done. could you please help with the workflow and merge? thanks

Comment (ileodo) [reply app/serializers/rest/instance_serializer.rb]: leave it to you to resolve the conversation.

Comment (ileodo) [reply app/serializers/rest/instance_serializer.rb]: @mjankowski could you please help to review it again and approve/merge if possible?

Comment (mjankowski) [reply app/serializers/rest/instance_serializer.rb]: I don't have merge permissions and I'm literally not qualified to review the JS portions of the change; but the spirit of things look correct to me and like an overall improvement, and I therefore wish you the absolute best of fates on your merge journey.

Comment (ileodo) [reply app/serializers/rest/instance_serializer.rb]: thanks. @renchap could you please help with?

**Comment by ileodo** [comment]
This PR has been here for ages.. can any one review and see if it can get merged??

**Comment by ClearlyClaire** [comment]
The authoring part is fine, but I'm not sure about the display part, for the following reasons:
1. limiting the number of media in a gallery client-side seems unnecessary if the server also enforces this limit (note that there is a caveat in that very old versions of Mastodon did not limit the number of attachments stored)
2. similarly, the request for the server configuration (`fetchServer()`) could be avoided
3. design-wise, the media gallery component does not gracefully handle more than 4 attachments: when more than 4 attachments are attached, only the first two will be displayed, and the remaining ones will be part of the DOM, but below the first two, overflowing the component and thus hidden; this probably needs bigger design changes for such a feature to be useful

**Comment by ileodo** [comment]
@ClearlyClaire I agree with what you said that there will be more work required if we want to support more than 4 media per post. This is why I made it clear in the PR title and description that the intention of this PR is just refactoring to have a constant indicate the limit instead having magic number 3/4 everywhere. 

Are we fine with the refactoring?

**Comment by ClearlyClaire** [comment]
Makes sense, but I think this would be easier done by:
- removing the `take(4)` from the front-end entirely
- possibly enforcing this server-side instead in `REST::StatusSerializer`

This would simplify the client-side code greatly and avoid doing an extra request to the API.

Further down the line, the UI can be changed to handle arbitrary many attachments in a more graceful way, and the limit can be lifted from the serializer.

**Comment by ileodo** [comment]
@ClearlyClaire  The client side need to be aware of this limit, because:
1. The WebUI currently have the logic to check the limit when user uploading medias.
2. Other clients (like iOS apps, e.g. IceCube) need to know the limit so that they can limit the same in the media picker widgets.

**Comment by ClearlyClaire** [comment]
> The client side need to be aware of this limit, because it's not only for display but also for uploading

I'm not disputing that, what I meant is that we can remove complexity from `app/javascript/mastodon/components/media_gallery.jsx` and `app/javascript/mastodon/containers/media_container.jsx` by not doing the checks here.

The changes in `app/javascript/mastodon/actions/compose.js` and `app/javascript/mastodon/features/compose/containers/upload_button_container.js` are fine.

**Comment by ileodo** [comment]
@ClearlyClaire updated as per your comments

**Comment by ileodo** [comment]
@ClearlyClaire may I know is it good to be merged now?

**Comment by ileodo** [comment]
can someone have a look at this PR? it has been there for ages...

**Comment by ileodo** [comment]
thanks, apart form #30932, anything you think we need for this PR?

**Review by ClearlyClaire — APPROVED** [review]

**Pull request merged**
